src/lib/InstantSearch.ts

```diff
@@ -243,6 +245,40 @@ See ${createDocumentationLink({
       const routerOptions = typeof routing === 'boolean' ? undefined : routing;
       this.EXPERIMENTAL_use(createRouter(routerOptions));
     }
+
+    if (this.insightsClient) {
+      this.setupUserTokenUpdater();
+    }
+  }
+
+  /**
+   * Updates userToken automatically.
+   *
+   * When insightsClient is given, this method sets up an event listener
+   * so that every time userToken is set to insightsClient, it can set it to
+   * searchParameter, too.
+   */
+  private setupUserTokenUpdater() {
+    const setUpdater = (callback: (userToken: string) => void) => {
+      (this.insightsClient as Function)('onSetUserToken', callback, {
+        immediate: true,
+      });
+    };
+
+    const customWidgetForUserToken = {
+      init({ helper }) {
+        setUpdater(userToken => {
+          const nextSearchParameters = mergeSearchParameters(
```

Comment:
there's no need for this, you can do directly `helper.setQueryParameter('userToken', userToken)` (even without widget IMHO)

Reply:
Ah right, we already have helper in the instance. I should use `this.helper` instead of `this.mainHelper`. Am I correct?

Reply:
Actually now that I think of it, it might be best to do it in a (fake) widget after all, but in the `getWidgetSearchParameters` part (as well as updating) to make sure it gets propagated to multi-index etc.

Reply:
I tried with `this.helper.setQueryParameter()` and it seems to work correctly with multi-index situations.